docs/versioning.md

```diff
@@ -0,0 +1,488 @@
+"Null Version Mapping".
+
+#### Specifying Versions in APIs for Putting Versions
+
+Since S3 does not currently allow overwrite of existing version data, Put
+Object, Complete Multipart Upload and Copy Object return `400 InvalidArgument`
+if a specific version ID is specified in the request query, e.g. for a
+`PUT /foo?versionId=v1` request.
+
+### PUT Example
+
+When S3 receives a request to PUT an object:
+
+- It checks first if versioning has been configured
+- If it has not been configured, S3 proceeds to delete any pre-existing data and
+  put the new object as the master version in Metadata
+
+If it has been configured, S3 checks the following:
+
+#### Versioning Enabled
+
+If versioning is enabled and there is existing object metadata:
+
+- If the master version is a null version or has no version ID (put before
+    versioning was configured):
+    - store the null version metadata as a new version
+    - create a new version and overwrite the master version
+        - set `nullVersionId`: version ID returned when storing null version
+
+If versioning is enabled and the master version is not null;
+or there is no existing object metadata:
+
+- create a new version and store it
+
+#### Versioning Suspended
+
+If versioning is suspended and there is existing object metadata:
+
+- If the master version is a null version or has no version ID:
+    - overwrite the master version with the new metadata
+    - delete previous object data
+- If master is not a null version and `nullVersionId` is defined in the object’s
+  metadata:
+    - delete the current null version metadata and data
+    - overwrite the master version with the new metadata
+
+If there is no existing object metadata, create the new null version as the
+master version.
+
+In each of the above cases, set `isNull` metadata attribute to true when when
+creating the new null version.
+
+### Behavior of Object-Targeting APIs
+
+API methods which can target existing objects or versions, such as Get Object,
+Head Object, Get Object ACL, Put Object ACL, Copy Object and Copy Part, will
+perform the action on the latest version of an object if no version ID is
+specified in the request query or relevant request header
+(`x-amz-copy-source-version-id` for Copy Object and Copy Part APIs).
+
+One exception are the Delete Object and Multi-Object Delete APIs, which will
+instead attempt to create delete markers, described in the following section, if
+no version ID is specified.
+
+No versioning options are necessary to retrieve the latest version from
+metadata, since the master version is stored in a key with the name of the
+object. However, when updating the latest version, such as with the Put Object
+ACL API, S3 sets the `versionId` option in the PUT call to metadata
+to the value stored in the object metadata's `versionId` attribute. This is done
+in order to update the metadata both in the master version and the version
+itself, if it is not a null version.
```

Comment:
Initially I assumed it would only update the master version. But I checked with Vinh, and it will create a new version with that versionId if it does not exist. We may need to add a check for this case in objectPutAcl and objectPutTagging.